### ISSUE TITLE
Bottom windows corner visual issue on some apps

### ISSUE THREAD

**Issue by bluebyt** [vinceliuice/Graphite-gtk-theme#118]
I have a visual issue at the bottom corners on some  apps like Firefox, totem, xfce4-terminal and more ...

![slurped](https://github.com/vinceliuice/Graphite-gtk-theme/assets/18442224/7c7b4c62-e802-4d87-b6e3-99d3e4ad1155)

**Comment by CoolAller** [comment]
Hi!
I think the best solution here would be to completely remove the window border, because there are GTK3 applications with edges rounded on all sides, and there are GTK3 applications that only have rounded edges on top and sharp bottom edges.

If you fix this problem for GTK3 applications with sharp bottom edges, then other GTK3 applications with rounded edges on all sides will not look the best.

This can be seen, for example, if you run the gnome-calculator application, which has all edges rounded, and the Celluloid application, which has sharp bottom edges.

In any case, you Need to fix these lines in the gtk.css (GTK3):
```diff
/**********************
 * Window Decorations *
 **********************/
decoration {
  transition: none;
  box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2), 0 10px 18px rgba(0, 0, 0, 0.14), 0 8px 24px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
  border-radius: 14px;
  border: 2px solid #E0E0E0;
  background-clip: border-box;
  margin: 10px;
}

decoration:backdrop {
  transition: box-shadow 75ms cubic-bezier(0, 0, 0.2, 1);
  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.2), 0 3px 4px rgba(0, 0, 0, 0.14), 0 1px 6px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
  border: 2px solid #E0E0E0;
  background-clip: border-box;
}
```
You can set the following values:
```diff
border-radius: 16px 16px 0px 0px;
```
And the celluloid application will look good, but the gnome-calculator application will have problems with the bottom corners, since they are made rounded in the application itself.

It seems to me that the best solution would be to completely remove the window border:
```diff
/**********************
 * Window Decorations *
 **********************/
decoration {
  transition: none;
  box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2), 0 10px 18px rgba(0, 0, 0, 0.14), 0 8px 24px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
  border-radius: 14px;
/* border-radius: 16px 16px 2px 2px;*/
-  border: 3px solid #E0E0E0;
  background-clip: border-box;
  margin: 10px;
}

decoration:backdrop {
  transition: box-shadow 75ms cubic-bezier(0, 0, 0.2, 1);
-  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.2), 0 3px 4px rgba(0, 0, 0, 0.14), 0 1px 6px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
+ box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2), 0 10px 18px rgba(0, 0, 0, 0.14), 0 1px 6px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
-  border: 3px solid #E0E0E0;
  background-clip: border-box;
}
```

**Comment by bluebyt** [comment]
I see, this is more a problem with Gnome applications then with the theme itself,  I will try both solution and see which one is better, thanks for the quick reply!

**Comment by swomf** [comment]
Depending on your use case, you might disable corner rounding and keep the white border. For example, KDE users here from `vinceliuice/Graphite-kde-theme` might only see the issue in Firefox.

![Firefox border with round corners](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/42f57e48-dbd8-4f8a-bc4f-5104296b4c19)

In this case, you can edit the following (I'm using the GTK theme `Graphite-dark-compact` so it is in `gtk-dark.css`)
-  `* Header bars *` ➔ `.titlebar {` 
-  `* Window Decorations *` ➔ `decoration {`

```diff
/***************
 * Header bars *                                          (after some scrolling)
 ***************/

 .titlebar {
   transition: background-color 75ms cubic-bezier(0, 0, 0.2, 1), color 75ms cubic-bezier(0, 0, 0.2, 1);
   background-color: #0F0F0F;
   color: white;
-  border-radius: 12px 12px 0 0;
+  border-radius: 0px 0px 0 0;
 }

/**********************
  * Window Decorations *
  **********************/
 decoration {
   transition: none;
   box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2), 0 10px 18px rgba(0, 0, 0, 0.14), 0 8px 24px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
-  border-radius: 14px;
   border: 2px solid #E0E0E0;
   background-clip: border-box;
   margin: 10px;
 }
```

Result:
![Firefox border without round corners](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/8bc4ca9e-2f5a-4b6d-9593-45baea61e8b7)

(Of course, KDE Firefox users can also use Customize Toolbar ➔ Title Bar)

**Comment by CoolAller** [comment]
[swomf](https://github.com/swomf)
It won't work that way. Gnome GTK-3 applications will have problems with interface corners.

<img src="https://github.com/vinceliuice/Graphite-gtk-theme/assets/9169898/145f0513-e2f1-4b8b-b175-5b362ef94b7a" width=70% height=70%>

**Comment by swomf** [comment]
What if, instead of using `border-radius` and `border` to outline borders, we use a `box-shadow`? I think other themes do that to have thin black borders on their windows.

**~/.themes/Graphite-Dark-compact/gtk-3.0/gtk-dark.css**

```diff
 /**********************
  * Window Decorations *
  **********************/
 decoration {
   transition: none;
-  box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2), 0 10px 18px rgba(0, 0, 0, 0.14), 0 8px 24px rgba(0, 0, 0, 0.12), 0 0 50
px transparent;
-  border-radius: 14px;
-  border: 2px solid #E0E0E0;
+  border-radius: 12px 12px 0 0;
+  border-width: 0px;
   background-clip: border-box;
+  box-shadow: 0 3px 12px 1px transparent, 0 2px 6px 2px rgba(255, 255, 255, 0.14), 0 0 0 2px shade(mix(white,white,
0.85),0.7);
   margin: 10px;
 }
 
  decoration:backdrop {
   transition: box-shadow 75ms cubic-bezier(0, 0, 0.2, 1);
-  box-shadow: 0 1px 2px rgba(0, 0, 0, 0.2), 0 3px 4px rgba(0, 0, 0, 0.14), 0 1px 6px rgba(0, 0, 0, 0.12), 0 0 50px transparent;
-  border: 2px solid #E0E0E0;
+  box-shadow: 0 3px 12px 1px transparent, 0 2px 6px 2px rgba(255, 255, 255, 0.14), 0 0 0 2px shade(mix(white,white,0.85),0.7);
   background-clip: border-box;
 }
```

Note that other `box-shadow` elements will have to be replaced accordingly, since we're sacrificing the old borders for a `box-shadow` border.

```diff
 /************
  * Chromium *
  ************/

 window.background.chromium decoration {
   background-clip: padding-box;
-  box-shadow: 0 3px 8px rgba(0, 0, 0, 0.2), 0 10px 18px rgba(0, 0, 0, 0.14), 0 8px 24px rgba(0, 0, 0, 0.12);
+  box-shadow: 0 3px 12px 1px transparent, 0 2px 6px 2px rgba(255, 255, 255, 0.14), 0 0 0 2px shade(mix(white,white,
0.85),0.7);
 }
```

Appearance:

![Firefox with box-shadow borders](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/d78028d1-9c45-475a-a6c1-272e3e3e6b76)

**Comment by CoolAller** [comment]
[swomf](https://github.com/swomf) Changing the order of addends does not change the sum, in our case - the result :)

If set border-radius: 12 12 0 0, then it is assumed that the bottom corners of the applications are sharp, but in the case of Gnome3 applications, they have rounded corners at the bottom. As a result, you will get the same result as in the screenshot above.

**Comment by swomf** [comment]
But I love the border [CoolAller](https://github.com/CoolAller)!

Buuut you're right. Different applications in different desktop environments will react differently to any tweaks we make, whether we want a round corner, or a not-round corner, *especially* when an application optimized for one DE is used in another DE. However, the `box-shadow` tweak *does* help with some DEs, especially in the major app Firefox. So it's worth consideration despite its assumption about borders.

For instance, when using the `box-shadow` hack on **Ubuntu 22.04 LTS (GNOME)**:
- XFCE Terminal (not GNOME) was unfixable with or without it
- non-snap Firefox couldn't be fixed by the tweak; I had to use Firefox's `about:config -> widget.gtk.rounded-bottom-corners.enabled` and restart the app
- I couldn't replicate the `gnome-calculator` issue. Hopefully it was fixed?

![ubuntu-xfceterm-notweak](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/858a9e6d-4ec6-4c5c-b83b-d3fdea7b632b)

When using the `box-shadow` hack in **ArchLinux KDE and also XFCE**:
- Firefox was fixed, with or without `about:config -> widget.gtk.rounded-bottom-corners.enabled`
- I couldn't find anything else, since gnome-calculator didn't want to use the theme at all.

![arch-xfce-posttweak](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/d2d9594c-5059-4e72-80c6-4ea55869ea0a)

When using the `box-shadow` hack in **Linux Mint Cinnamon Edition**:
- Firefox is fixed, but only *without* using the bottom corner rounding option in `about:config` (different from GNOME earlier)
- Chromium gets glitched around the bookmarks bar area
- I couldn't find anything else
**before**
![lm-cinnamon-pretweak-1](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/5147096d-d669-42eb-bb6e-ffe3d7d48bfc)
![lm-cinnamon-pretweak-2](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/f8dbf662-be6b-4f31-830c-bd1efa4675dc)
**after**
![lm-cinnamon-posttweak-1](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/153a06fd-be57-4b6d-b39b-b144a1561a32)
![lm-cinnamon-posttweak-2](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/a21d0d00-96a6-4a29-b94b-3775e04d942f)
Enabling bottom corner rounding in Firefox here will reveal the issue you predicted of my assumption. 
![cinnamon-firefox-bottom-round-glitch](https://github.com/vinceliuice/Graphite-gtk-theme/assets/108753551/4c7d969f-449c-4ee3-a7be-358fc03fbdcf)

In summary, I think the `box-shadow` edit could help people keep the pretty GTK border. ^_^

**Comment by CoolAller** [comment]
[swomf](https://github.com/swomf) The main thing is that you are satisfied with how your Desktop looks. Anyway, in the case of Linux Desktop Environments, you always have to make some compromises. 
To be honest, I don’t use this theme, I just answered the question from [bluebyt](https://github.com/bluebyt).  :)
